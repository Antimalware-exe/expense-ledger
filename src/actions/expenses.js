import uuid from 'uuid';
//  Add expense generator
export const addExpense = (
  { description = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
      id: uuid(),
      description,
      note,
      amount,
      createdAt
    }
  }
);

// Delete expense generator
export const deleteExpense = ({ id }) => ({
  type: 'DELETE_EXPENSE',
  id
});

//Edit expense generator
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});