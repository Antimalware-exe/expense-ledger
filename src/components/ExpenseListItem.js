import React from 'react';
// import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
// import { deleteExpense } from '../actions/expenses';
import moment from 'moment'

const ExpenseListItem = ({ /*dispatch,*/ id, description, amount, createdAt }) => (
  <div>
    <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
    <span>Amount: {amount}, Created at: {moment(createdAt).format('DD/MM/YYYY')}</span>
    {/*
      <button id={id} onClick={() => {
      dispatch(deleteExpense({ id }));
      }}>Delete</button>
    */}
  </div>
);

export default ExpenseListItem;