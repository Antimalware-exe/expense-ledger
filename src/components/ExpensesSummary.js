import React from 'react';
import { connect } from 'react-redux';
import expensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export class ExpensesSummary extends React.Component {
  render() {
    return (
      <div>
        <h3>
          <p>Expense count: {this.props.expenses.length}</p>
          <p>Total expense: â‚¹{expensesTotal(this.props.expenses)}</p>
        </h3>
      </div>
    )
  };
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);