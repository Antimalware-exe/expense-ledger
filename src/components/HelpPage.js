import React from 'react';

const HelpPage = () => (
    <div>Display help</div>
)

export default HelpPage;