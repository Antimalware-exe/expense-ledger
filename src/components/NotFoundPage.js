import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <div>
        404! Page not found. - <Link to="/">Go to Home</Link>
    </div>
);

export default NotFoundPage;