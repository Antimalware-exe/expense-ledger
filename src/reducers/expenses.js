//Expenses Reducer
const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      // return state.concat(action.expense);  // concat method returns new array and does not modifies state array
      return [...state, action.expense];
    case 'DELETE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return { ...expense, ...action.updates };
        } else return expense;
      });
    default:
      return state;
  }
};