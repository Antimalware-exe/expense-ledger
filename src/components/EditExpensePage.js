import React from 'react';
import { connect } from 'react-redux';
import { editExpense, startDeleteExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
    onSubmit = (update) => {
        this.props.editExpense(this.props.expense.id, update);
        this.props.history.push('/'); //redirects to the mentioned url
    };

    onClick = () => {
        this.props.startDeleteExpense({ id: this.props.expense.id });
        this.props.history.push('/'); //redirects to the mentioned url
    };

    render() {
        return (
            <div>
                <h3>Edit Expense</h3>
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
                <button onClick={this.onClick}>Delete</button>
            </div>
        );
    };
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id, update) => dispatch(editExpense(id, update)),
        startDeleteExpense: (id) => dispatch(startDeleteExpense(id))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);