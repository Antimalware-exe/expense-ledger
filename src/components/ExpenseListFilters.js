import React from 'react';
import { DateRangePicker } from 'react-dates';
import { connect } from 'react-redux';
import { changeText, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = (e) => {
    this.props.changeText(e.target.value);
  };

  onSortChange = (e) => {
    e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input className="text-input" type="text" placeholder="search"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select className="select" onChange={this.onSortChange}>
              <option value="date"> Sort by date</option>
              <option value="amount">Sort by amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              startDateId="someStartid" // required id in latest update
              endDateId="someEndId"
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>



      </div>
    )
  };
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  changeText: (text) => dispatch(changeText(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);