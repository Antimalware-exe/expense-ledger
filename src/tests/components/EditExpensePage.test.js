import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense, deleteExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  deleteExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage expense={expenses[1]} editExpense={editExpense} deleteExpense={deleteExpense} history={history} />);
});

test('should render editExpense page correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  const update = {
    note: 'New note added'
  };
  wrapper.find('ExpenseForm').prop('onSubmit')(update);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, update);
});

test('should handle onClick', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(deleteExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
});