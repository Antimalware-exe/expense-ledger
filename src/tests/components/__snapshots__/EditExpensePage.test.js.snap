// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render editExpense page correctly 1`] = `
<div>
  <h3>
    Edit Expense
  </h3>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
